src/python/pants/jvm/goals/coursier.py

```diff
@@ -20,28 +21,85 @@
 )
 from pants.engine.goal import Goal, GoalSubsystem
 from pants.engine.rules import Get, MultiGet, collect_rules, goal_rule, rule
-from pants.engine.target import Sources, Target, Targets
-from pants.jvm.resolve.coursier_fetch import ArtifactRequirements, CoursierResolvedLockfile
-from pants.jvm.target_types import JvmLockfileSources, MavenRequirementsField
+from pants.engine.target import InvalidTargetException, Sources, Target, Targets, UnexpandedTargets
+from pants.jvm.resolve.coursier_fetch import (
+    ArtifactRequirements,
+    Coordinate,
+    CoursierResolvedLockfile,
+)
+from pants.jvm.target_types import (
+    JvmArtifactArtifactField,
+    JvmArtifactGroupField,
+    JvmArtifactVersionField,
+    JvmLockfileSources,
+    JvmRequirementsField,
+)
+from pants.util.logging import LogLevel
 
 
 class CoursierResolveSubsystem(GoalSubsystem):
     name = "coursier-resolve"
-    help = "Generate a lockfile by resolving Maven dependencies."
+    help = "Generate a lockfile by resolving JVM dependencies."
 
 
 class CoursierResolve(Goal):
     subsystem_cls = CoursierResolveSubsystem
 
 
+@dataclass(frozen=True)
+class GatherArtifactRequirementsRequest:
+    """Gather the coordinate requirements from a JarRequirementsField."""
+
+    requirements: JvmRequirementsField
+
+
+@rule(level=LogLevel.DEBUG)
+async def gather_artifact_requirements(
+    request: GatherArtifactRequirementsRequest,
+) -> ArtifactRequirements:
+    def from_target(tgt: Target) -> Coordinate:
+        group = tgt[JvmArtifactGroupField].value
+        if not group:
+            raise InvalidTargetException(
+                f"The `group` field of {tgt.alias} target {tgt.address} must be set."
+            )
+
+        artifact = tgt[JvmArtifactArtifactField].value
+        if not artifact:
+            raise InvalidTargetException(
+                f"The `artifact` field of {tgt.alias} target {tgt.address} must be set."
+            )
+
+        version = tgt[JvmArtifactVersionField].value
+        if not version:
+            raise InvalidTargetException(
+                f"The `version` field of {tgt.alias} target {tgt.address} must be set."
+            )
+
+        return Coordinate(
+            group=group,
+            artifact=artifact,
+            version=version,
+        )
+
+    requirements_addresses = await Get(
+        Addresses, UnparsedAddressInputs, request.requirements.to_unparsed_address_inputs()
+    )
+    requirements_targets = await Get(UnexpandedTargets, Addresses, requirements_addresses)
```

Comment:
FYI the meaning of `UnexpandedTargets` vs. `Targets` has changed. Now, it means: "should you apply target generation?" And that target generation is generic, like if you ever add a target generator for `jvm_artifact`.

https://github.com/pantsbuild/pants/blob/f03f96f158eaa45c95fc59b94e70a94164138e92/src/python/pants/engine/internals/graph.py#L152-L181

So, you should probably use `Targets` here to future-proof any target generators that generate `jvm_artifact`.

Reply:
Will switch to `Targets` then.